File: python/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        buy = prices[0]

        for price in prices:
            if price < buy:
                buy = price
            else:
                maxSoFar = price - buy
                maxProfit = max(maxProfit, maxSoFar)


        return maxProfit

File: javascript/121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
  let maxProfit = 0;
  let buy = prices[0];

  for(let i = 1; i < prices.length; i++) {
      let maxSoFar = prices[i] - buy;

      if(maxSoFar < 0) {
          buy = prices[i]
      } else {
          maxProfit = Math.max(maxProfit, maxSoFar);
      }
  }

  return maxProfit;
};